### PR TITLE
[#187] Check globs in config fields and CLI args

### LINKED ISSUE #187: Forbid absolute paths in config fields
# Clarification and motivation
Original discussion [#152](https://github.com/serokell/xrefcheck/pull/152#discussion_r983634482)

Currenly when we see absolute filepaths (e.g. `/file1.md)` in virtualFiles/ignored/notScanned fields of config, we interpret them as absolute filepaths **in filesystem**, which confusing (since in markdown they are absolute from repo root) and is not useful and leads to non-reproducibility. We could treat them as absolute from repo root, but in this case there will be no difference between `file1.md`
 and `/file1.md`, so I think we should just forbid such paths (throwing readable message)

# Acceptance criteria
We throw error when see absolute filepath in config fields

### PR DESCRIPTION
**Pull request by Sorokin-Anton**
## Description
Problem: user can think that globs in e.g. config field `ignored` have same root as all markdown links, so if he will add e.g. `/scripts/*` to ignore, link `[a](/scripts/*)` will be not checked. But we were understanding this glob as filesystem-level absolute path (so this link will still be checked). Also we pretend to check correctness of user-supplied globs, but we're using glob parser in mode that always succeed (and transform bad globs to correct somehow).

Solution:  make function `mkGlobPattern` that throws readable error when see absolute paths (like `/scripts/*`) or malformed globs (like `<a>`), use it for parsing all globs
in config fields and cli args. Also remove redundant "canonization" of globs

## Related issue(s)

Fixes #187

## :white_check_mark: Checklist for your Pull Request

Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](https://github.com/serokell/xrefcheck/tree/master/README.md)
    - Haddock

- Public contracts
  - [ ] Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - [ ] I added an entry to the [changelog](https://github.com/serokell/xrefcheck/tree/master/CHANGES.md) if my changes are visible to the users
        and
  - [ ] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](https://github.com/serokell/style/blob/master/haskell.md).

#### ✓ Release Checklist

- [ ] I updated the version number in `package.yaml`.
- [ ] I updated the [changelog](https://github.com/serokell/xrefcheck/tree/master/CHANGES.md) and moved everything
      under the "Unreleased" section to a new section for this release version.
- [ ] (After merging) I edited the [auto-release](https://github.com/serokell/xrefcheck/releases/tag/auto-release).
    * Change the tag and title using the format `vX.Y.Z`.
    * Write a summary of all user-facing changes.
    * Deselect the "This is a pre-release" checkbox at the bottom.
- [ ] (After merging) I updated [`xrefcheck-action`](https://github.com/serokell/xrefcheck-action#updating-supported-versions).
- [ ] (After merging) I uploaded the package to [hackage](https://hackage.haskell.org/package/xrefcheck).

### CODE REVIEW

**Review by Sereja313 — APPROVED** [review]

File: src/Xrefcheck/System.hs [hunk 1]
```diff
@@ -37,9 +39,18 @@ askWithinCI = lookupEnv "CI" <&> \case
...
 
--- | Glob pattern relative to repository root.
+-- | Glob pattern relative to repository root. Should be created via @mkGlobPattern@
 newtype RelGlobPattern = RelGlobPattern FilePath
 
+mkGlobPattern :: ToString s => s -> Either String RelGlobPattern
+mkGlobPattern path = do
+  let spath = toString path
+  unless (isRelative spath) $ Left $
+    "Expected a relative glob, but got " <> spath
```
Comment (dcastro) [inline src/Xrefcheck/System.hs:49]:
```suggestion
    "Expected a relative glob pattern, but got " <> spath
```

File: tests/golden/check-ignored/check-ignored.bats [hunk 2]
```diff
@@ -39,3 +39,31 @@ load '../helpers'
...
+    --ignored "/to-ignore/*"
+  assert_failure
+  assert_output --partial "option --ignored: Expected a relative glob, but got /to-ignore/*"
+}
+
+@test "CLI: Malformed glob in arg \"ignored\" yields to error" {
+  run xrefcheck\
+    --ignored "<to-ignore>"
+  assert_failure
+  assert_output --partial "option --ignored: compile :: bad <>"
```
Comment (dcastro) [inline tests/golden/check-ignored/check-ignored.bats:68]: I guess we could enrich the error message here a bit. 

When we fail to compile a glob pattern, we could tell the user to consult the syntax for glob patterns here: https://hackage.haskell.org/package/Glob/docs/System-FilePath-Glob.html#v:compile

---

We could also tell the user how to escape characters (as far as I can tell, this isn't documented anywhere). I found [a trick here](https://stackoverflow.com/a/2595162/857807).

Basically, when you have a file literally named `<a>.md` and you want to ignore it, `xrefcheck --ignored "<a>.md"` will fail to compile the pattern.

But if you wrap special characters in `[` and `]`, it'll work as expected: `xrefcheck --ignored "[<]a[>].md"
`

Comment (Sorokin-Anton) [reply tests/golden/check-ignored/check-ignored.bats]: Done

**Review by dcastro — APPROVED** [review]
